internal/graph/check.go

```diff
@@ -0,0 +1,558 @@
+		close(subChan)
+	}()
+
+	baseHandler := handlers[0]
+	subHandler := handlers[1]
+
+	limiter <- struct{}{}
+	wg.Add(1)
+	go func() {
+		resp, err := baseHandler(ctx)
+		baseChan <- checkOutcome{resp, err}
+		<-limiter
+		wg.Done()
+	}()
+
+	limiter <- struct{}{}
+	wg.Add(1)
+	go func() {
+		resp, err := subHandler(ctx)
+		subChan <- checkOutcome{resp, err}
+		<-limiter
+		wg.Done()
+	}()
+
+	for i := 0; i < len(handlers); i++ {
+		select {
+		case baseResult := <-baseChan:
+			if baseResult.err != nil {
+				return &openfgapb.CheckResponse{Allowed: false}, baseResult.err
+			}
+
+			if !baseResult.resp.Allowed {
+				return &openfgapb.CheckResponse{Allowed: false}, nil
+			}
+		case subResult := <-subChan:
+			if subResult.err != nil {
+				return &openfgapb.CheckResponse{Allowed: false}, subResult.err
+			}
+
+			if subResult.resp.Allowed {
+				return &openfgapb.CheckResponse{Allowed: false}, nil
+			}
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+
+	return &openfgapb.CheckResponse{Allowed: true}, nil
+}
+
+// dispatch dispatches the provided Check request to the CheckDispatcher this LocalChecker
+// was constructed with.
+func (c *LocalChecker) dispatch(ctx context.Context, req *dispatcher.DispatchCheckRequest) CheckHandlerFunc {
+	return func(ctx context.Context) (*openfgapb.CheckResponse, error) {
+		resp, err := c.dispatcher.DispatchCheck(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+
+		return &openfgapb.CheckResponse{
+			Allowed: resp.Allowed,
+		}, nil
+	}
+}
+
+func (c *LocalChecker) DispatchCheck(
+	ctx context.Context,
+	req *dispatcher.DispatchCheckRequest,
+) (*dispatcher.DispatchCheckResponse, error) {
+	ctx, span := tracer.Start(ctx, "DispatchCheck")
+	defer span.End()
+
+	if req.GetResolutionMetadata().Depth == 0 {
+		return nil, ErrResolutionDepthExceeded
+	}
+
+	typesys, ok := typesystem.TypesystemFromContext(ctx)
+	if !ok {
+		panic("typesystem missing in context")
+	}
+
+	object := req.GetTupleKey().GetObject()
+	relation := req.GetTupleKey().GetRelation()
+
+	objectType, _ := tuple.SplitObject(object)
+	rel, err := typesys.GetRelation(objectType, relation)
+	if err != nil {
+		return nil, fmt.Errorf("relation '%s' undefined for object type '%s'", relation, objectType)
+	}
+
+	resp, err := union(ctx, c.concurrencyLimit, c.checkRewrite(ctx, req, rel.GetRewrite()))
+	if err != nil {
+		return nil, err
+	}
+
+	return &dispatcher.DispatchCheckResponse{
+		Allowed: resp.Allowed,
+	}, nil
+}
+
+// checkDirect composes two CheckHandlerFunc which evaluate direct relationships with the provided
+// 'object#relation'. The first handler looks up direct matches on the provided 'object#relation@user',
+// while the second handler looks up relationships between the target 'object#relation' and any usersets
+// related to it.
+func (c *LocalChecker) checkDirect(parentctx context.Context, req *dispatcher.DispatchCheckRequest) CheckHandlerFunc {
+
+	typesys, ok := typesystem.TypesystemFromContext(parentctx)
```

Comment:
The parameter `parentctx` is used here and then never again, can we pass the `typesys` directly? same comment for `checkTTU`

Reply:
I'm favoring propagating the typesystem via the context because I want us to start propagating the typesystem in the context. Other refactors will come along with this change. If we've resolved the authorization model somewhere up the call chain, then there should be a typesystem in the context. I don't want to have to pass down the TypeSystem reference everywhere in the call chain that needs it.

Reply:
I don't have a strong opinion on this either way. Is there any documentation of best practices for what to store in the context? From what I read so far it seems like it is a nice place to store request-specific data there. In this sense perhaps it does makes sense the store the storeID and modelID or typesystem in there. It would make our signatures simpler, but then we also lose some explicitness... I need to do some reading.

Reply:
Guidance I've read on Go's official blog and doc is to use context for request scoped data. 

Here's my rationale for the way I did this.. Contextual tuples seemed like a good fit to propagate down to the storage layer to remove the responsibility of handling tuples in the request context differently than the other tuples coming from the storage layer. It's always been a struggle for me to look at our current code and see this special handling of contextual tuples. To me they are just an "extension" of the tuples returned from the database, and they should be treated that way. Our Check, Expand, and ListObjects code should operate on tuples/iterators returned from the storage layer and not be bothered with contextual specific tuples coming in the request context. I believe that responsibility should be saved for the storage layer which is providing tuple filtering and querying capabilities for us already.